File: src/processing/mode/android/AndroidRunner.java
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
 Part of the Processing project - http://processing.org

 Copyright (c) 2012-17 The Processing Foundation
 Copyright (c) 2011-12 Ben Fry and Casey Reas

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License version 2
 as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software Foundation,
 Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

package processing.mode.android;

import java.io.PrintStream;
import java.security.PublicKey;
import java.util.Iterator;
import java.util.List;
import java.util.concurrent.*;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

import jdk.nashorn.internal.runtime.linker.Bootstrap;
import processing.app.ui.Editor;
import processing.app.Messages;
import processing.app.RunnerListener;
import processing.app.SketchException;
import processing.mode.java.runner.Runner;

/** 
 * Launches an app on the device or in the emulator.
 */
public class AndroidRunner implements DeviceListener {
  AndroidBuild build;
  RunnerListener listener;

  protected PrintStream sketchErr;
  protected PrintStream sketchOut;

  public boolean debugEnabled;

  public AndroidRunner(AndroidBuild build, RunnerListener listener) {
    this.build = build;
    this.listener = listener;

    if (listener instanceof Editor) {
      Editor editor = (Editor) listener;
      sketchErr = editor.getConsole().getErr();
      sketchOut = editor.getConsole().getOut();
    } else {
      sketchErr = System.err;
      sketchOut = System.out;
    }
  }


  public boolean launch(Future<Device> deviceFuture, int comp, boolean emu) {
    String devStr = emu ? "emulator" : "device";
    listener.statusNotice("Waiting for " + devStr + " to become available...");
    
    final Device device = waitForDevice(deviceFuture, listener);
    if (device == null || !device.isAlive()) {
      listener.statusError("Lost connection with " + devStr + " while launching. Try again.");
      // Reset the server, in case that's the problem. Sometimes when
      // launching the emulator times out, the device list refuses to update.
      final Devices devices = Devices.getInstance();
      devices.killAdbServer();
      return false;
    }
    
    if (comp == AndroidBuild.WATCHFACE && !device.hasFeature("watch")) {
      listener.statusError("Could not install the sketch.");
      Messages.showWarning("Selected device is not a watch...",
          "You are trying to install a watch face on a non-watch device.\n" +
          "Select the correct device, or use the emulator.");      
      return false;
    }
    
    if (comp != AndroidBuild.WATCHFACE && device.hasFeature("watch")) {
      listener.statusError("Could not install the sketch.");
      Messages.showWarning("Selected device is a watch...",
          "You are trying to install a non-watch app on a watch.\n" +
          "Select the correct device, or use the emulator.");      
      return false;
    }

    device.addListener(this);
    device.setPackageName(build.getPackageName());
    device.setSketchClassName(build.getSketchClassName());
    device.debugEnabled = debugEnabled;
    listener.statusNotice("Installing sketch on " + device.getId());
    // this stopped working with Android SDK tools revision 17
    if (!device.installApp(build, listener)) {
      listener.statusError("Lost connection with " + devStr + " while installing. Try again.");
      final Devices devices = Devices.getInstance();
      devices.killAdbServer();  // see above
      return false;
    }

    boolean status = false;
    if (comp == AndroidBuild.WATCHFACE || comp == AndroidBuild.WALLPAPER) {
      if (startSketch(build, device)) {
        listener.statusNotice("Sketch installed "
                              + (device.isEmulator() ? "in the emulator" : "on the device") + ".");
        status = true;
      } else {
        listener.statusError("Could not install the sketch.");
      }
    } else {
      listener.statusNotice("Starting sketch on " + device.getId());
      if (startSketch(build, device)) {
        listener.statusNotice("Sketch launched "
                              + (device.isEmulator() ? "in the emulator" : "on the device") + ".");
        status = true;
      } else {
        listener.statusError("Could not start the sketch.");
      }
    }
    
    listener.stopIndeterminate();
    lastRunDevice = device;
    return status;
  }


  private volatile Device lastRunDevice = null;


  // if user asks for 480x320, 320x480, 854x480 etc, then launch like that
  // though would need to query the emulator to see if it can do that

  private boolean startSketch(AndroidBuild build, final Device device) {
    final String packageName = build.getPackageName();
    try {
      if (device.launchApp(packageName)) {
        return true;
      }
    } catch (final Exception e) {
      e.printStackTrace(System.err);
    }
    return false;
  }


  private Device waitForDevice(Future<Device> deviceFuture, RunnerListener listener) {
    for (int i = 0; i < 120; i++) {
      if (listener.isHalted()) {
        deviceFuture.cancel(true);
        return null;
      }
      try {
        return deviceFuture.get(1, TimeUnit.SECONDS);
      } catch (final InterruptedException e) {
        listener.statusError("Interrupted.");
        return null;
      } catch (final ExecutionException e) {
        listener.statusError(e);
        return null;
      } catch (final TimeoutException expected) {
      }
    }
    listener.statusError("No, on second thought, I'm giving up " +
                         "on waiting for that device to show up.");
    return null;
  }


  private static final Pattern LOCATION =
    Pattern.compile("\\(([^:]+):(\\d+)\\)");
  private static final Pattern EXCEPTION_PARSER =
    Pattern.compile("^\\s*([a-z]+(?:\\.[a-z]+)+)(?:: .+)?$",
                    Pattern.CASE_INSENSITIVE);

  /**
   * Currently figures out the first relevant stack trace line
   * by looking for the telltale presence of "processing.android"
   * in the package. If the packaging for droid sketches changes,
   * this method will have to change too.
   */
  public void stackTrace(final List<String> trace) {
    final Iterator<String> frames = trace.iterator();
    final String exceptionLine = frames.next();

    final Matcher m = EXCEPTION_PARSER.matcher(exceptionLine);
    if (!m.matches()) {
      System.err.println("Can't parse this exception line:");
      System.err.println(exceptionLine);
      listener.statusError("Unknown exception");
      return;
    }
    final String exceptionClass = m.group(1);
    Runner.handleCommonErrors(exceptionClass, exceptionLine, listener, sketchErr);

    while (frames.hasNext()) {
      final String line = frames.next();
      if (line.contains("processing.android")) {
        final Matcher lm = LOCATION.matcher(line);
        if (lm.find()) {
          final String filename = lm.group(1);
          final int lineNumber = Integer.parseInt(lm.group(2)) - 1;
          final SketchException rex =
            build.placeException(exceptionLine, filename, lineNumber);
          listener.statusError(rex == null ? new SketchException(exceptionLine, false) : rex);
          return;
        }
      }
    }
  }


  // called by AndroidMode.handleStop()...
  public void close() {
    if (lastRunDevice != null) {
      lastRunDevice.bringLauncherToFront();
    }
  }


  // sketch stopped on the device
  public void sketchStopped() {
    listener.stopIndeterminate();
    listener.statusHalt();
  }
}

File: mode/src/processing/mode/android/AndroidRunner.kt
/* -*- mode: java; c-basic-offset: 2; indent-tabs-mode: nil -*- */

/*
 Part of the Processing project - http://processing.org

 Copyright (c) 2012-17 The Processing Foundation
 Copyright (c) 2011-12 Ben Fry and Casey Reas

 This program is free software; you can redistribute it and/or modify
 it under the terms of the GNU General Public License version 2
 as published by the Free Software Foundation.

 This program is distributed in the hope that it will be useful,
 but WITHOUT ANY WARRANTY; without even the implied warranty of
 MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 GNU General Public License for more details.

 You should have received a copy of the GNU General Public License
 along with this program; if not, write to the Free Software Foundation,
 Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */
package processing.mode.android

import com.sun.jdi.VMDisconnectedException
import com.sun.jdi.VirtualMachine
import com.sun.jdi.connect.AttachingConnector
import com.sun.jdi.connect.IllegalConnectorArgumentsException

import org.eclipse.jdi.Bootstrap

import processing.app.Messages
import processing.app.RunnerListener
import processing.app.SketchException
import processing.app.ui.Editor
import processing.mode.android.AndroidMode.Companion.getTextString
import processing.mode.java.runner.Runner

import java.io.IOException
import java.io.PrintStream
import java.util.concurrent.ExecutionException
import java.util.concurrent.Future
import java.util.concurrent.TimeUnit
import java.util.concurrent.TimeoutException
import java.util.regex.Pattern

/**
 * Launches an app on the device or in the emulator.
 */
internal class AndroidRunner(var build: AndroidBuild, var listener: RunnerListener) : DeviceListener {
    private var sketchErr: PrintStream? = null
    private var sketchOut: PrintStream? = null
    private var vm: VirtualMachine? = null
    private var isDebugEnabled = false

    fun launch(deviceFuture: Future<Device>, comp: Int, emu: Boolean): Boolean {
        val devStr = if (emu) "emulator" else "device"
        listener.statusNotice(getTextString("android_runner.status.waiting_for_device", devStr))
        val device = waitForDevice(deviceFuture, listener)
        if (device == null || !device.isAlive) {
            listener.statusError(getTextString("android_runner.status.lost_connection_with_device", devStr))
            // Reset the server, in case that's the problem. Sometimes when
            // launching the emulator times out, the device list refuses to update.
            val devices = Devices.getInstance()
            devices.killAdbServer()
            return false
        }
        if (comp == AndroidBuild.WATCHFACE && !device.hasFeature("watch")) {
            listener.statusError(getTextString("android_runner.status.cannot_install_sketch"))
            Messages.showWarning(getTextString("android_runner.warn.non_watch_device_title"),
                    getTextString("android_runner.warn.non_watch_device_body"))
            return false
        }
        if (comp != AndroidBuild.WATCHFACE && device.hasFeature("watch")) {
            listener.statusError(getTextString("android_runner.status.cannot_install_sketch"))
            Messages.showWarning(getTextString("android_runner.warn.watch_device_title"),
                    getTextString("android_runner.warn.watch_device_body"))
            return false
        }
        device.addListener(this)
        device.setPackageName(build.packageName)
        listener.statusNotice(getTextString("android_runner.status.installing_sketch", device.id))
        // this stopped working with Android SDK tools revision 17
        if (!device.installApp(build, listener)) {
            listener.statusError(getTextString("android_runner.status.lost_connection", devStr))
            val devices = Devices.getInstance()
            devices.killAdbServer() // see above
            return false
        }
        var status = false
        if (comp == AndroidBuild.WATCHFACE || comp == AndroidBuild.WALLPAPER) {
            if (startSketch(build, device)) {
                listener.statusNotice(getTextString("android_runner.status.sketch_installed")
                        + (if (device.isEmulator) " " + getTextString("android_runner.status.in_emulator") else " " +
                        getTextString("android_runner.status.on_device")) + ".")
                status = true
            } else {
                listener.statusError(getTextString("android_runner.status.cannot_install_sketch"))
            }
        } else {
            listener.statusNotice(getTextString("android_runner.status.launching_sketch", device.id))
            if (startSketch(build, device)) {
                listener.statusNotice(getTextString("android_runner.status.sketch_launched")
                        + (if (device.isEmulator) " " + getTextString("android_runner.status.in_emulator") else " " +
                        getTextString("android_runner.status.on_device")) + ".")
                status = true
            } else {
                listener.statusError(getTextString("android_runner.status.cannot_launch_sketch"))
            }
        }

        // Start Debug if Debugger is enabled
        if (isDebugEnabled) {
            (listener as AndroidEditor).debugger
                    .startDebug(this, device)
        }
        listener.stopIndeterminate()
        lastRunDevice = device
        return status
    }

    @Throws(IOException::class)
    fun connectVirtualMachine(port: Int): VirtualMachine? {
        val strPort = Integer.toString(port)
        val connector = connector

        return try {
            vm = connect(connector, strPort)
            vm
        } catch (e: IllegalConnectorArgumentsException) {
            throw IllegalStateException(e)
        }
    }

    private val connector: AttachingConnector
        private get() {
            val vmManager = Bootstrap.virtualMachineManager()
            for (connector in vmManager.attachingConnectors()) {
                if ("com.sun.jdi.SocketAttach" == connector.name()) {
                    return connector as AttachingConnector
                }
            }
            throw IllegalStateException()
        }

    @Throws(IllegalConnectorArgumentsException::class, IOException::class)
    private fun connect(
            connector: AttachingConnector, port: String): VirtualMachine {
        val args = connector
                .defaultArguments()

        val pidArgument = args["port"] ?: throw IllegalStateException()
        pidArgument.setValue(port)

        return connector.attach(args)
    }

    fun vm(): VirtualMachine? {
        return vm
    }

    @Volatile
    private var lastRunDevice: Device? = null

    // if user asks for 480x320, 320x480, 854x480 etc, then launch like that
    // though would need to query the emulator to see if it can do that
    private fun startSketch(build: AndroidBuild, device: Device): Boolean {
        val packageName = build.packageName

        try {
            if (device.launchApp(packageName, isDebugEnabled)) {
                return true
            }
        } catch (e: Exception) {
            e.printStackTrace(System.err)
        }

        return false
    }

    private fun waitForDevice(deviceFuture: Future<Device>, listener: RunnerListener): Device? {
        for (i in 0..119) {
            if (listener.isHalted) {
                deviceFuture.cancel(true)
                return null
            }

            try {
                return deviceFuture[1, TimeUnit.SECONDS]
            } catch (e: InterruptedException) {
                listener.statusError("Interrupted.")
                return null
            } catch (e: ExecutionException) {
                listener.statusError(e)
                return null
            } catch (expected: TimeoutException) {
            }
        }

        listener.statusError(getTextString("android_runner.status.cancel_waiting_for_device"))

        return null
    }

    /**
     * Currently figures out the first relevant stack trace line
     * by looking for the telltale presence of "processing.android"
     * in the package. If the packaging for droid sketches changes,
     * this method will have to change too.
     */
    override fun stackTrace(trace: List<String?>?) {
        val frames = trace!!.iterator()
        val exceptionLine = frames.next()
        val m = EXCEPTION_PARSER.matcher(exceptionLine)

        if (!m.matches()) {
            System.err.println(getTextString("android_runner.error.cannot_parse_stacktrace"))
            System.err.println(exceptionLine)
            listener.statusError(getTextString("android_runner.status.unknwon_exception"))
            return
        }

        val exceptionClass = m.group(1)

        Runner.handleCommonErrors(exceptionClass, exceptionLine, listener, sketchErr)

        while (frames.hasNext()) {
            val line = frames.next()
            if (line!!.contains(DEFAULT_PACKAGE_NAME)) {
                val lm = LOCATION.matcher(line)
                if (lm.find()) {
                    val filename = lm.group(1)
                    val lineNumber = lm.group(2).toInt() - 1
                    val rex = build.placeException(exceptionLine, filename, lineNumber)
                    listener.statusError(rex ?: SketchException(exceptionLine, false))
                    return
                }
            }
        }
    }

    // called by AndroidMode.handleStop()...
    fun close() {
        if (lastRunDevice != null) {
            lastRunDevice!!.bringLauncherToFront()
        }

        if (vm != null) {
            try {
                vm!!.exit(0)
            } catch (vmde: VMDisconnectedException) {
                // if the vm has disconnected on its own, ignore message
                //System.out.println("harmless disconnect " + vmde.getMessage());
                // TODO shouldn't need to do this, need to do more cleanup
            }
        }
    }

    // sketch stopped on the device
    override fun sketchStopped() {
        listener.stopIndeterminate()
        listener.statusHalt()
    }

    companion object {

        private const val DEFAULT_PACKAGE_NAME = "processing.android"

        private val LOCATION = Pattern.compile("\\(([^:]+):(\\d+)\\)")

        private val EXCEPTION_PARSER = Pattern.compile("^\\s*([a-z]+(?:\\.[a-z]+)+)(?:: .+)?$",
                Pattern.CASE_INSENSITIVE)
    }

    // constructor or initializer
    init {
        if (listener is AndroidEditor) {
            isDebugEnabled = (listener as AndroidEditor).isDebuggerEnabled
        }
        if (listener is Editor) {
            val editor = listener as Editor
            sketchErr = editor.console.err
            sketchOut = editor.console.out
        } else {
            sketchErr = System.err
            sketchOut = System.out
        }
    }
}